File: tests/browser/offline.mjs
// `offline` in a real Chrome, against `tests/fixtures/offline`, built twice —
// pre-rendered and as a single page — each driven through what a reader
// meets: the worker stores the site; a stored route loads with the server
// gone and one that is not shows the fallback; a write made then is counted,
// kept and sent once when the server returns; a new build is offered and
// taken with one reload, and the old version's store goes.
//
//   cd tests/browser && npm ci && node offline.mjs
//
// WF (default target/release/wf) and CHROME (default /usr/bin/google-chrome)
// say where they are. Service workers need a real browser: none of this can
// be run against the fake DOM the other suites use.

import puppeteer from "puppeteer-core";
import { execFileSync } from "node:child_process";
import fs from "node:fs";
import http from "node:http";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

const repo = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
const WF = process.env.WF || path.join(repo, "target/release/wf");
const CHROME = process.env.CHROME || "/usr/bin/google-chrome";
const fixture = path.join(repo, "tests/fixtures/offline");

const TYPES = { ".html": "text/html", ".js": "text/javascript", ".css": "text/css", ".json": "application/json" };

/// The build served as a static host would, and `POST /api/save` recorded.
function host(root, port) {
  const received = [];
  const server = http.createServer((req, res) => {
    if (req.method === "POST") {
      let body = "";
      req.on("data", (c) => { body += c; });
      req.on("end", () => {
        received.push(`${req.url} ${body}`);
        res.writeHead(200, { "Content-Type": "application/json" }).end('{"ok":true}');
      });
      return;
    }
    let file = path.join(root, decodeURIComponent(new URL(req.url, "http://x").pathname));
    if (fs.existsSync(file) && fs.statSync(file).isDirectory()) file = path.join(file, "index.html");
    if (!fs.existsSync(file)) {
      // A single-page build answers every route with its shell.
      const shell = path.join(root, "index.html");
      if (path.extname(file) === "" && fs.existsSync(shell)) file = shell;
      else return res.writeHead(404).end("not found");
    }
    res.writeHead(200, { "Content-Type": TYPES[path.extname(file)] || "application/octet-stream" });
    fs.createReadStream(file).pipe(res);
  });
  return {
    received,
    start: () => new Promise((r) => server.listen(port, "127.0.0.1", r)),
    stop: () => new Promise((r) => { server.closeAllConnections?.(); server.close(() => r()); }),
  };
}

const results = [];
function check(name, ok, detail = "") {
  results.push(ok);
  console.log(`  ${ok ? "PASS" : "FAIL"}  ${name}${detail && !ok ? `  (${detail})` : ""}`);
}
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const until = (page, fn, arg, timeout = 15000) =>
  page.waitForFunction(fn, { timeout }, arg).then(() => true, () => false);

async function scenario(label, ssg, port) {
  console.log(`\n${label}`);
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "wf-offline-site-"));
  fs.cpSync(fixture, dir, { recursive: true, filter: (p) => !p.includes(`${path.sep}build`) });
  const configPath = path.join(dir, "webfluent.app.json");
  const config = JSON.parse(fs.readFileSync(configPath, "utf8"));
  config.build.ssg = ssg;
  fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
  execFileSync(WF, ["build", "-d", dir], { stdio: "ignore" });
  const build = path.join(dir, "build");
  const worker = fs.readFileSync(path.join(build, "sw.js"), "utf8");
  const stores = JSON.parse(/const C = (\{.*?\});\n/.exec(worker)[1]).precache.slice().sort();

  const site = host(build, port);
  await site.start();
  const origin = `http://localhost:${port}`;
  const profile = fs.mkdtempSync(path.join(os.tmpdir(), "wf-offline-chrome-"));
  const browser = await puppeteer.launch({ executablePath: CHROME, headless: true, userDataDir: profile, args: ["--no-sandbox"] });
  const page = await browser.newPage();
  let loads = 0;
  page.on("load", () => { loads += 1; });
  try {
    await page.goto(`${origin}/`, { waitUntil: "load" });
    check("the worker controls the page after the first visit",
      await until(page, () => navigator.serviceWorker.controller !== null));
    const stored = await page.evaluate(async () => {
      const store = (await caches.keys()).find((k) => /^wf-[0-9a-f]{16}$/.test(k));
      return store ? (await (await caches.open(store)).keys()).map((r) => new URL(r.url).pathname).sort() : [];
    });
    check("it stored everything the worker names", JSON.stringify(stored) === JSON.stringify(stores), stored.join(" "));

    await site.stop();
    await page.goto(`${origin}/about/`, { waitUntil: "load" });
    check("a stored route loads with the server gone",
      await until(page, () => document.querySelector("h1")?.textContent === "About Us"),
      await page.evaluate(() => document.querySelector("h1")?.textContent));
    await page.goto(`${origin}/contact/`, { waitUntil: "load" });
    check("a route not stored shows the fallback",
      await until(page, () => document.querySelector("h1")?.textContent === "You are offline"),
      await page.evaluate(() => `${location.pathname} ${document.querySelector("h1")?.textContent}`));

    await page.goto(`${origin}/`, { waitUntil: "load" });
    await page.waitForSelector("#save");
    await page.click("#save");
    check("a write made offline is kept, and network.queued counts it",
      await until(page, () => document.getElementById("queued")?.textContent === "queued:1", undefined, 5000));
    check("the action goes on as if it were sent",
      await until(page, () => document.getElementById("saved")?.textContent === "saved:sent or kept", undefined, 5000));

    await site.start();
    await page.evaluate(() => window.dispatchEvent(new Event("online")));
    check("it is sent when the server returns",
      await until(page, () => document.getElementById("queued")?.textContent === "queued:0"));
    await sleep(1500);
    check("the server received it once, as written",
      site.received.length === 1 && site.received[0].includes("written while offline"), JSON.stringify(site.received));

    const home = path.join(dir, "src/App.wf");
    fs.writeFileSync(home, fs.readFileSync(home, "utf8").replace('Heading("Offline test")', 'Heading("Offline test, v2")'));
    execFileSync(WF, ["build", "-d", dir], { stdio: "ignore" });
    await page.goto(`${origin}/`, { waitUntil: "load" });
    await page.evaluate(async () => (await navigator.serviceWorker.getRegistration()).update());
    check("a new build is offered: update.available", await until(page, () => !!document.getElementById("update")));
    const before = loads;
    await page.click("#update");
    check("update.apply() takes it, and the page shows it",
      await until(page, () => document.querySelector("h1")?.textContent === "Offline test, v2"));
    await sleep(2500);
    check("the page reloaded once, not twice", loads - before === 1, `${loads - before} loads`);
    const versions = await page.evaluate(async () => (await caches.keys()).filter((k) => /^wf-[0-9a-f]{16}$/.test(k)));
    check("the old version's store is gone", versions.length === 1, versions.join(", "));
  } catch (e) {
    check(`the run (${e.message})`, false);
  } finally {
    await browser.close();
    await site.stop().catch(() => {});
    fs.rmSync(profile, { recursive: true, force: true });
    fs.rmSync(dir, { recursive: true, force: true });
  }
}

await scenario("Pre-rendered (build.ssg)", true, 4371);
await scenario("A single page", false, 4372);
const failed = results.filter((ok) => !ok).length;
console.log(`\n${results.length - failed} of ${results.length} passed`);
process.exit(failed ? 1 : 0);

File: tests/offline.rs
//! What `offline` in the config makes a build write, and refuse.
//!
//! How the worker behaves — storing, serving with the server gone, a write
//! kept and sent, the update flow — is `tests/browser/offline.mjs`, in a real
//! Chrome, since a service worker is nothing without one. This holds what
//! the build decides: what `sw.js` stores and routes, its version, where the
//! bundle registers it, and the configurations it refuses.

use std::path::{Path, PathBuf};
use std::process::Command;

fn scratch(name: &str, config: &str, app: &str) -> PathBuf {
    let dir = std::env::temp_dir().join(format!("wf-offline-{name}-{}", std::process::id()));
    let _ = std::fs::remove_dir_all(&dir);
    std::fs::create_dir_all(dir.join("src")).unwrap();
    std::fs::write(dir.join("webfluent.app.json"), config).unwrap();
    std::fs::write(dir.join("src/App.wf"), app).unwrap();
    dir
}

fn build(dir: &Path) -> (bool, String) {
    let out = Command::new(env!("CARGO_BIN_EXE_wf"))
        .args(["build", "-d"])
        .arg(dir)
        .output()
        .expect("run wf");
    (
        out.status.success(),
        format!(
            "{}{}",
            String::from_utf8_lossy(&out.stdout),
            String::from_utf8_lossy(&out.stderr)
        ),
    )
}

/// The JSON the worker was built with.
fn worker_config(dir: &Path) -> serde_json::Value {
    let sw = std::fs::read_to_string(dir.join("build/sw.js")).expect("sw.js");
    let start = sw.find("const C = ").unwrap() + "const C = ".len();
    let end = start + sw[start..].find(";\n").unwrap();
    serde_json::from_str(&sw[start..end]).unwrap()
}

const PAGES: &str = r#"
app { Router }
page Home(path: "/", title: "Home", description: "d") { Heading("Home").h1 }
page About(path: "/about", title: "About", description: "d") { Heading("About").h1 }
page Contact(path: "/contact", title: "Contact", description: "d") { Heading("Contact").h1 }
page Offline(path: "/offline", title: "Offline", description: "d") { Heading("Offline").h1 }
"#;

fn config(ssg: bool, offline: &str, base: &str) -> String {
    format!(
        r#"{{ "name": "t", "build": {{ "output": "build", "ssg": {ssg}, "base_path": "{base}" }}, "offline": {offline} }}"#
    )
}

#[test]
fn a_static_build_stores_the_shell_the_named_routes_and_the_fallback() {
    let dir = scratch(
        "ssg",
        &config(
            true,
            r#"{ "precache": ["/", "/about"], "fallback": "/offline", "cache": { "/api/*": "network-first" }, "sync": true }"#,
            "",
        ),
        PAGES,
    );
    let (ok, out) = build(&dir);
    assert!(ok, "{out}");
    let c = worker_config(&dir);
    let mut stored: Vec<&str> = c["precache"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();
    stored.sort();
    assert_eq!(
        stored,
        [
            "/about/index.html",
            "/app.js",
            "/index.html",
            "/offline/index.html",
            "/pages/About.js",
            "/pages/Home.js",
            "/pages/Offline.js",
            "/styles.css"
        ],
        "a route the config does not name — /contact — is not stored"
    );
    assert_eq!(
        c["fallback"], "/offline",
        "the fallback is a route to send a navigation to"
    );
    assert_eq!(c["sync"], true);
    assert_eq!(
        c["policies"][0],
        serde_json::json!(["^/api/.*$", "network-first"])
    );
    let routes: Vec<String> = c["routes"]
        .as_array()
        .unwrap()
        .iter()
        .map(|r| r[0].as_str().unwrap().to_string())
        .collect();
    assert!(
        routes.contains(&"/about".to_string()) && routes.contains(&"/offline".to_string()),
        "{routes:?}"
    );

    let bundle = std::fs::read_to_string(dir.join("build/app.js")).unwrap();
    assert!(
        bundle.contains("WF.offline({sync:true})"),
        "the bundle registers the worker"
    );
    assert!(
        bundle.contains("wf:activate"),
        "and carries the module that does it"
    );
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn a_single_page_build_stores_the_shell_and_answers_its_routes_with_it() {
    let dir = scratch(
        "spa",
        &config(false, r#"{ "precache": ["/about"] }"#, ""),
        PAGES,
    );
    let (ok, out) = build(&dir);
    assert!(ok, "{out}");
    let c = worker_config(&dir);
    assert_eq!(c["shell"], "/index.html");
    assert!(
        c["precache"]
            .as_array()
            .unwrap()
            .iter()
            .any(|v| v == "/pages/About.js")
    );
    assert!(
        !c["precache"]
            .as_array()
            .unwrap()
            .iter()
            .any(|v| v == "/pages/Contact.js")
    );
    // A stored route is answered by the shell; the worker has to know which
    // those are, or it would send them to the fallback like the rest.
    assert_eq!(c["routes"], serde_json::json!([["/about", "/index.html"]]));
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn every_path_is_under_the_base() {
    let dir = scratch(
        "base",
        &config(true, r#"{ "fallback": "/offline" }"#, "/site"),
        PAGES,
    );
    let (ok, out) = build(&dir);
    assert!(ok, "{out}");
    let c = worker_config(&dir);
    assert_eq!(c["base"], "/site");
    for path in c["precache"].as_array().unwrap() {
        assert!(path.as_str().unwrap().starts_with("/site/"), "{path}");
    }
    assert_eq!(c["fallback"], "/site/offline");
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn the_version_moves_with_the_build_and_only_then() {
    let dir = scratch("version", &config(true, "{}", ""), PAGES);
    assert!(build(&dir).0);
    let first = worker_config(&dir)["version"].as_str().unwrap().to_string();
    assert!(build(&dir).0);
    assert_eq!(
        worker_config(&dir)["version"],
        first.as_str(),
        "nothing changed, so neither does the version"
    );
    std::fs::write(
        dir.join("src/App.wf"),
        PAGES.replace("Heading(\"About\")", "Heading(\"About us\")"),
    )
    .unwrap();
    assert!(build(&dir).0);
    assert_ne!(
        worker_config(&dir)["version"],
        first.as_str(),
        "a changed page is a new version"
    );
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn without_offline_nothing_is_registered_or_written() {
    let dir = scratch(
        "none",
        r#"{ "name": "t", "build": { "output": "build" } }"#,
        PAGES,
    );
    assert!(build(&dir).0);
    assert!(!dir.join("build/sw.js").exists());
    let bundle = std::fs::read_to_string(dir.join("build/app.js")).unwrap();
    assert!(!bundle.contains("WF.offline"), "no registration");
    assert!(!bundle.contains("wf:activate"), "and not the module either");
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn what_offline_cannot_mean_is_refused() {
    let dir = scratch(
        "refused",
        &config(
            true,
            r#"{ "fallback": "/gone", "cache": { "/api/*": "network-last" } }"#,
            "",
        ),
        PAGES,
    );
    let (ok, out) = build(&dir);
    assert!(!ok);
    assert!(
        out.contains("`network-last`") && out.contains("`network-first`"),
        "{out}"
    );
    assert!(out.contains("`/gone`, which no page's `path` is"), "{out}");
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn update_and_network_queued_are_values_a_page_reads() {
    let dir = scratch(
        "values",
        &config(true, r#"{ "sync": true }"#, ""),
        r#"
page Home(path: "/", title: "Home", description: "d") {
    Heading("Home").h1
    Text("waiting: {network.queued}")
    if update.available { Button("Update") { on click { update.apply() } } }
}
"#,
    );
    let (ok, out) = build(&dir);
    assert!(ok, "the type checker knows both: {out}");
    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn the_host_is_told_never_to_cache_the_worker() {
    let dir = scratch(
        "headers",
        r#"{ "name": "t", "build": { "output": "build", "ssg": true, "csp": true, "base_path": "/site" }, "offline": {} }"#,
        PAGES,
    );
    let (ok, out) = build(&dir);
    assert!(ok, "{out}");
    let headers = std::fs::read_to_string(dir.join("build/_headers")).unwrap();
    assert!(
        headers.contains("/site/sw.js\n  Cache-Control: no-cache"),
        "{headers}"
    );
    let _ = std::fs::remove_dir_all(&dir);
}
